### PR TITLE
Share four exception types

### PR DESCRIPTION
**Pull request by danmosemsft**
Share four easy exceptions.

Temporarily adding a failfast in __Error to try to see whether tests hit any cases we're dropping the file path.

### CODE REVIEW

**Comment by danmosemsft** [comment]
@dotnet-bot test Windows_NT x86 corefx_baseline
@dotnet-bot test Ubuntu x64 corefx_baseline

**Comment by danmosemsft** [comment]
Test failure was https://github.com/dotnet/coreclr/issues/13486
I'll remove the  temp code then.

**Comment by danmosemsft** [comment]
@dotnet-bot test Windows_NT x86 corefx_baseline

**Comment by danmosemsft** [comment]
@dotnet-bot help please

**Comment by danmosemsft** [comment]
CoreFX test failures are unrelated - https://github.com/dotnet/corefx/issues/22791 and #13486. CI is clean.

File: src/mscorlib/shared/System.Private.CoreLib.Shared.projitems [hunk 1]
```diff
@@ -322,6 +325,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\ReferenceAssemblyAttribute.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\RuntimeCompatibilityAttribute.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\RuntimeFeature.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\RuntimeWrappedException.cs" />
```
Comment (jkotas) [inline src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: - The CoreRT version of RuntimeWrappedException.cs has proper formatting (no `m_` prefix, fields first). Can we keep the formatting from the CoreRT copy.
- The CoreRT version has one private constructor public to workaround private reflection issues. This needs to be done under ifdef in the shared copy like in other places. It would break .NET Native tests otherwise.

File: src/mscorlib/System.Private.CoreLib.csproj [hunk 2]
```diff
@@ -276,7 +275,6 @@
     <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\WindowsRuntime\CLRIKeyValuePairImpl.cs" />
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="$(BclSourcesRoot)\System\AggregateException.cs" />
```
Comment (jkotas) [inline src/mscorlib/System.Private.CoreLib.csproj]: The name of the serializable `m_innerExceptions` field is different between CoreCLR and CoreRT. It should have been detected by the binary serialization tests. Do we have a test hole in corefx for this?

File: src/mscorlib/shared/System.Private.CoreLib.Shared.projitems [hunk 3]
```diff
@@ -19,7 +19,9 @@
     <Compile Include="$(MSBuildThisFileDirectory)Microsoft\Win32\SafeHandles\SafeHandleMinusOneIsInvalid.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Microsoft\Win32\SafeHandles\SafeHandleZeroOrMinusOneIsInvalid.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Action.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\AccessViolationException.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\ApplicationException.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\AggregateException.cs" />
```
Comment (jkotas) [inline src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: CoreRT copy of AggregateException has argument null check removed in GetObjectData because of it is done by the base class. Same change should be done in CoreCLR too.

Comment (jkotas) [inline src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: The implementation of `AggregateException.ToString` in CoreRT looks better: 
- Correct handling of localized strings (CoreCLR version is mixing localizable `>>>` with hardcoded `<<<` 
- Less allocations (CoreCLR version will allocate `O(N^2)` where N is number of inner exceptions)

File: src/mscorlib/src/System/IO/__Error.cs [hunk 4]
```diff
@@ -120,24 +121,27 @@ internal static void WinIOError(int errorCode, String str)
                     throw new DriveNotFoundException(SR.Format(SR.IO_DriveNotFound_Drive, str));
 
                 case Win32Native.ERROR_INVALID_PARAMETER:
-                    throw new IOException(Win32Native.GetMessage(errorCode), Win32Native.MakeHRFromErrorCode(errorCode), str);
+                    Debug.Assert(String.IsNullOrEmpty(str), "message should include path");
```
Comment (jkotas) [inline src/mscorlib/src/System/IO/__Error.cs]: Do you plan to keep these asserts?

Note that there is almost identical code in GetExceptionForWin32Error in the shared partition that does not have asserts like this.

File: src/mscorlib/System.Private.CoreLib.csproj (hunk 2 above)
Comment (danmosemsft) [inline src/mscorlib/System.Private.CoreLib.csproj]: @ViktorHofer it seems the tests don't store serialized blobs for Exception and AggregateException specifically. If that's the case it would not detect issues serializing them between runtimes, I think? Do you recall why that is? 
```c#
        public static IEnumerable<object> SerializableExceptions_MemberData()
        {
            var exception = new Exception("Exception message", new Exception("Inner exception message"));
            yield return new object[] { new AggregateException("Aggregate exception message", exception) };
            yield return new object[] { exception };
        }
```

Comment (ViktorHofer) [inline src/mscorlib/System.Private.CoreLib.csproj]: I think we forgot to add blobs for exception types. I recall that in an offline thread we also got info from the ILC team that the serialized StackTrace might not be exchangeable between ILC and Core and Desktop. We at least should at test blobs for the two Exception types.

Comment (danmosemsft) [inline src/mscorlib/System.Private.CoreLib.csproj]: Took offline in https://github.com/dotnet/corefx/issues/23520

File: src/mscorlib/shared/System/IO/Win32Marshal.cs [hunk 5]
```diff
@@ -65,8 +65,11 @@ internal static Exception GetExceptionForWin32Error(int errorCode, string path)
                 case Interop.Errors.ERROR_FILENAME_EXCED_RANGE:
                     return new PathTooLongException(SR.Format(SR.IO_PathTooLong_Path, path));
 
+                case Interop.Errors.ERROR_INVALID_DRIVE:
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/IO/Win32Marshal.cs]: This was missing in the __Error copy and it's possible this could cause some exceptions to now be DriveNotFoundException that were previously IOException. I figure that's OK as it's a derived type.

File: src/mscorlib/shared/System.Private.CoreLib.Shared.projitems (hunk 1 above)
Comment (danmosemsft) [inline src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: fixed

File: src/mscorlib/shared/System.Private.CoreLib.Shared.projitems (hunk 3 above)
Comment (danmosemsft) [inline src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: fixed

Comment (danmosemsft) [inline src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: fixed

**Comment by danmosemsft** [comment]
@dotnet-bot test Windows_NT x86 corefx_baseline
@dotnet-bot test Ubuntu x64 corefx_baseline

File: src/mscorlib/shared/System.Private.CoreLib.Shared.projitems (hunk 1 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: Turns out RWE needs it:
`Compiler optimization assumption invalid: EE expects field to exist: System.Runtime.CompilerServices.RuntimeWrappedException:m_wrappedException`

**Comment by danmosemsft** [comment]
@dotnet-bot test Windows_NT x86 corefx_baseline
@dotnet-bot test Ubuntu x64 corefx_baseline

File: src/mscorlib/shared/System.Private.CoreLib.Shared.projitems (hunk 1 above)
Comment (jkotas) [reply src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: It should be trivial to rename on the EE side (src\vm\mscorlib.h).

File: src/mscorlib/shared/System/IO/Win32Marshal.cs (hunk 5 above)
Comment (jkotas) [reply src/mscorlib/shared/System/IO/Win32Marshal.cs]: There is exact same core in CoreFX as well. You may want to fix it there as well.

File: src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs [hunk 6]
```diff
@@ -2,41 +2,39 @@
...
 {
+    /// <summary>
+    /// Exception used to wrap all non-CLS compliant exceptions.
+    /// </summary>
     public sealed class RuntimeWrappedException : Exception
     {
+        private Object m_wrappedException; // EE expects this name
+
+#if CORECLR
         private RuntimeWrappedException(Object thrownObject)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs]: Nit: All existing places have the ifdef around the modifier only, e.g.: https://github.com/dotnet/coreclr/blob/26fbb1933b1ead4168f271ab4ced0fd9906adbe0/src/mscorlib/shared/System/Collections/ListDictionaryInternal.cs#L24

File: src/mscorlib/shared/System.Private.CoreLib.Shared.projitems [hunk 7]
```diff
@@ -582,6 +582,12 @@
     <Compile Include="$(MSBuildThisFileDirectory)Interop\Windows\FileApiInterop\Interop.CreateFile2FromApp.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Interop\Windows\Kernel32\Interop.CREATEFILE2_EXTENDED_PARAMETERS.cs" />
   </ItemGroup>
+  <ItemGroup Condition="$(TargetsWindows) or '$(FeaturePal)'=='true'">
+    <Compile Include="$(MSBuildThisFileDirectory)Interop\Windows\Interop.Errors.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Interop\Windows\Interop.Libraries.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Interop\Windows\Kernel32\Interop.FormatMessage.cs" />
```
Comment (jkotas) [inline src/mscorlib/shared/System.Private.CoreLib.Shared.projitems]: I think there will be more ifdefs required to make this work ... we can wait for the CI results.

File: src/mscorlib/shared/System/IO/Win32Marshal.cs (hunk 5 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/IO/Win32Marshal.cs]: will do

File: src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs [hunk 8]
```diff
@@ -2,41 +2,40 @@
...
 
 namespace System.Runtime.CompilerServices
 {
+    /// <summary>
+    /// Exception used to wrap all non-CLS compliant exceptions.
+    /// </summary>
     public sealed class RuntimeWrappedException : Exception
     {
-        private RuntimeWrappedException(Object thrownObject)
+        private Object wrappedException; // EE expects this name
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs]: Add underscore prefix to match coding conventions

File: src/mscorlib/shared/System/IO/DriveNotFoundException.cs [hunk 9]
```diff
@@ -2,38 +2,34 @@
...
+            HResult = HResults.COR_E_DIRECTORYNOTFOUND;
         }
 
         protected DriveNotFoundException(SerializationInfo info, StreamingContext context) : base(info, context)
         {
             throw new PlatformNotSupportedException();
         }
     }
-}
+}
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/DriveNotFoundException.cs]: Nit: Add end of line

File: src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs [hunk 10]
```diff
@@ -2,41 +2,40 @@
...
+
+#if CORECLR
+        private
+#else
+        // Not an api but has to be public as System.Linq.Expression invokes this through Reflection when an expression
+        // throws an object that doesn't derive from Exception.
+        public
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs]: On a second thought, it may be best to make this public unconditionally and fix Systerm.Linq.Expressions to create it directly, and not using private reflection hack.

Comment (danmosemsft) [reply src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs]: That will need API review .. I can do that subsequently.

**Comment by danmosemsft** [comment]
@dotnet-bot test Ubuntu16.04 arm Cross Debug Build (out of disk space)

File: src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs (hunk 10 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs]: API review is required to expose it in contracts. I agree that it would be nice.

System.Linq.Expression can call it without exposing it in contracts. Flipping it to public here does not require API review.

**Comment by jkotas** [comment]
@dotnet-bot test Ubuntu x64 corefx_baseline

**Comment by jkotas** [comment]
CoreFX tests are failing on this because of the public DriveNotFoundException. Could you pleade flip it to internal temporarily to get successful CoreFX run?

**Comment by danmosemsft** [comment]
@dotnet-bot test Windows_NT x86 corefx_baseline
@dotnet-bot test Ubuntu x64 corefx_baseline

**Comment by danmosemsft** [comment]
The ubuntu failure is existing issue https://github.com/dotnet/corefx/issues/23535

**Comment by danmosemsft** [comment]
@jkotas good to merge?

**Comment by jkotas** [comment]
I am surprised that the change around Win32Marshal works on Unix. I want to take a closer look.

**Comment by danmosemsft** [comment]
Just rebased.

File: src/mscorlib/shared/System/IO/Win32Marshal.cs [hunk 11]
```diff
@@ -84,7 +90,7 @@ internal static Exception GetExceptionForWin32Error(int errorCode, string path)
                     return new OperationCanceledException();
 
                 default:
-                    return new IOException(GetMessage(errorCode), MakeHRFromErrorCode(errorCode));
+                    return new IOException(Interop.Kernel32.GetMessage(errorCode), MakeHRFromErrorCode(errorCode));
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/Win32Marshal.cs]: If you exercise this codepath by calling `System.IO.Win32Marshal.GetExceptionForWin32Error(1234567);`, it will fail with:

```
jkotas-mac:OSX.x64.Debug jkotas$ ./corerun test.exe 

Unhandled Exception: System.DllNotFoundException: Unable to load DLL 'kernel32.dll': The specified module or one of its dependencies could not be found.
 (Exception from HRESULT: 0x8007007E)
   at Interop.Kernel32.FormatMessage(Int32 dwFlags, IntPtr lpSource, UInt32 dwMessageId, Int32 dwLanguageId, StringBuilder lpBuffer, Int32 nSize, IntPtr[] arguments)
   at Interop.Kernel32.TryGetErrorMessage(IntPtr moduleHandle, Int32 errorCode, StringBuilder sb, String& errorMsg)
   at Interop.Kernel32.GetMessage(IntPtr moduleHandle, Int32 errorCode)
   at Interop.Kernel32.GetMessage(Int32 errorCode)
   at System.IO.Win32Marshal.GetExceptionForWin32Error(Int32 errorCode, String path)
   at System.IO.Win32Marshal.GetExceptionForWin32Error(Int32 errorCode)
```

Apparently, we do not have any coverage in our tests for this codepath on Unix. But there may be some rare situations where it will end up being called. I would be useful to not break it.

The options I can think of:
1. Dumb it down on Unix so that it does not call Kernel32. Make it a message with error code or something.
2. Or fix the ifdefs so that the shared code still reroutes the Kernel32 calls into the CoreCLR PAL.

Either one is fine with me.

File: src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs [hunk 12]
```diff
@@ -2,41 +2,35 @@
...
+        public override void GetObjectData(SerializationInfo info, StreamingContext context)
         {
-            get { return m_wrappedException; }
+            base.GetObjectData(info, context);
         }
 
-        private Object m_wrappedException;
-
-        public override void GetObjectData(SerializationInfo info, StreamingContext context)
+        public Object WrappedException
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/CompilerServices/RuntimeWrappedException.cs]: Nit: This can be: ` public Object WrappedException => _wrappedException;`.

**Comment by danmosemsft** [comment]
@dotnet/dnceng the ubuntu machine seems broken:
```
22:27:05 Building remotely on ubuntu1404-20170821-1854e0 (ubuntu1404-20170821) in workspace /mnt/j/workspace/dotnet_coreclr/master/checked_ubuntu_prtest
22:27:05 [WS-CLEANUP] Deleting project workspace...
22:27:05 java.io.IOException: Failed to mkdirs: /mnt/j/workspace/dotnet_coreclr/master/checked_ubuntu_prtest
22:27:05 	at hudson.FilePath.mkdirs(FilePath.java:1191)
22:27:05 	at hudson.model.AbstractProject.checkout(AbstractProject.java:1276)
22:27:05 	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:604)
22:27:05 	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
22:27:05 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:529)
22:27:05 	at hudson.model.Run.execute(Run.java:1720)
22:27:05 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
22:27:05 	at hudson.model.ResourceController.execute(ResourceController.java:98)
22:27:05 	at hudson.model.Executor.run(Executor.java:404)
```
https://ci.dot.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_flow_prtest/10429/

@dotnet-bot test Ubuntu x64 Checked Build and Test  please

**Comment by mmitche** [comment]
@danmosemsft This was probably an Azure issue mounting the disk.  The startup script should detect that the VM is unusable.  I'll file an issue for it.

**Comment by mmitche** [comment]
@danmosemsft Yeah it was just one machine.  I'll add a check in the startup script that the drive is writable.

**Comment by danmosemsft** [comment]
@dotnet-bot test ubuntu x64 build and test

**Comment by danmosemsft** [comment]
@dotnet-bot test ubuntu x64 checked build and test

**Comment by danmosemsft** [comment]
Ubuntu is https://github.com/dotnet/coreclr/issues/13736.
